alternator/ttl.cc

```diff
@@ -113,4 +139,321 @@ future<executor::request_return_type> executor::describe_time_to_live(client_sta
+// understands it is an expiration event - not a user-initiated deletion.
+static future<> expire_item(service::storage_proxy& proxy,
+                            const service::query_state& qs,
+                            const std::vector<bytes_opt>& row,
+                            const std::vector<unsigned>& pk_columns,
+                            const std::vector<unsigned>& ck_columns,
+                            schema_ptr schema,
+                            api::timestamp_type ts) {
+    // Prepare the row key to delete
+    std::vector<bytes> exploded_pk;
+    for (unsigned c : pk_columns) {
+        const auto& row_c = row[c];
+        if (!row_c) {
+            // This shouldn't happen - all key columns must have values.
+            // But if it ever happens, let's just *not* expire the item.
+            return make_ready_future<>();
+        }
+        exploded_pk.push_back(*row_c);
+    }
+    auto pk = partition_key::from_exploded(exploded_pk);
+    mutation m(schema, pk);
+    // If there's no clustering key, a tombstone should be created directly
+    // on a partition, not on a clustering row - otherwise it will look like
+    // an open-ended range tombstone, which will crash on KA/LA sstable format.
+    // See issue #6035
+    if (ck_columns.empty()) {
+        m.partition().apply(tombstone(ts, gc_clock::now()));
+    } else {
+        std::vector<bytes> exploded_ck;
+        for (unsigned c : ck_columns) {
+            const auto& row_c = row[c];
+            if (!row_c) {
+                // This shouldn't happen - all key columns must have values.
+                // But if it ever happens, let's just *not* expire the item.
+                return make_ready_future<>();
+            }
+            exploded_ck.push_back(*row_c);
+        }
+        auto ck = clustering_key::from_exploded(exploded_ck);
+        m.partition().clustered_row(*schema, ck).apply(tombstone(ts, gc_clock::now()));
+    }
+    return proxy.mutate(std::vector<mutation>{std::move(m)},
+        db::consistency_level::LOCAL_QUORUM,
+        executor::default_timeout(), // FIXME - which timeout?
+        qs.get_trace_state(), qs.get_permit());
+}
+
+
+future<> expiration_service::run() {
+    // FIXME: don't just tight-loop, think about timing, pace, and
+    // store position in durable storage, etc.
+    // FIXME: If one of the iterations below fails with an exception,
+    // this service will die forever. It's probably not a good idea!
+    // We should log the failure, but not stop the service.
+    for (;;) {
+        // FIXME: think about working on different tables in parallel.
+        // also need to notice when a new table is added, a table is
+        // deleted or when ttl is enabled or disabled for a table!
+        for (const auto& cf : _db.get_column_families()) {
+            schema_ptr s = cf.second->schema();
+            // Check if an expiration-time attribute is enabled for this
+            // table.
```

Comment:
It will be nicer to extract a single table's expiration into its own function.

Reply:
The reason I didn't do this yet is that I wasn't sure how I was going to organize this code in the long run. Will I have the individual tables being expired in parallel, or one expiration thread going over a piece of one table now, then a piece of a second table, and so on and so on.

Maybe I should have a function that scans a single table starting at a specific position - and then I can use this function either for scanning an entire table (as the code does today) or in the future for scanning a piece of a table starting in some position. I'll try to see how it works out.

Reply:
By the way, In the 12th patch I did introduce an inner function - not for an entire table but a specific range in a table. So I'll mark this resolved.